python/pyspark/sql/dataframe.py

```diff
@@ -1830,6 +1830,32 @@ def agg(self, *exprs):
         """
         return self.groupBy().agg(*exprs)
 
+    @since(3.3)
+    def observe(self, observation, *exprs):
+        """
+        Observe (named) metrics through an :class:`Observation` instance.
+        This method does not support streaming datasets.
+
+        .. versionadded:: 3.3.0
+
+        A user can retrieve the metrics by accessing `Observation.get`.
+
+        Example:
+            >>> from pyspark.sql.functions import col, lit, count, max
+            >>> from pyspark.sql.observation import Observation
+            >>> observation = Observation("my_metrics")
+            >>> observed_df = df.observe(observation, count(lit(1)).as("rows"), max(col("id")).as("maxid"))
+            >>> observed_df.write.parquet("ds.parquet")
+            >>> metrics = observation.get
+
+        :param observation: :class:`Observation` instance
+        :param exprs: aggregation expressions
+        :return: observed :class:`DataFrame`
```

Comment:
Can you follow NumPy documentation style? (see also https://spark.apache.org/docs/latest/api/python/development/contributing.html#code-and-docstring-guide)

Reply:
Done